lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1234,6 +1234,79 @@ class BuildControl : public FuncOpPartialLoweringPattern {
   }
 };
 
+/// This pass recursively inlines use-def chains of combinational logic (from
+/// non-stateful groups) into groups referenced in the control schedule.
+class InlineCombGroups
+    : public PartialLoweringPattern<calyx::GroupInterface,
+                                    OpInterfaceRewritePattern> {
+public:
+  InlineCombGroups(MLIRContext *context, LogicalResult &resRef,
+                   ProgramLoweringState &pls)
+      : PartialLoweringPattern(context, resRef), pls(pls) {}
+
+  LogicalResult partiallyLower(calyx::GroupInterface originGroup,
+                               PatternRewriter &rewriter) const override {
+    auto &state = pls.compLoweringState(
+        originGroup->getParentOfType<calyx::ComponentOp>());
+
+    /// Filter groups which are not part of the control schedule.
+    if (SymbolTable::symbolKnownUseEmpty(originGroup.symName(),
+                                         state.getComponentOp().getControlOp()))
+      return success();
+
+    /// Maintain a set of the groups which we've inlined so far. The group
+    /// itself is implicitly inlined.
+    llvm::SmallSetVector<Operation *, 8> inlinedGroups;
+    inlinedGroups.insert(originGroup);
+
+    /// Starting from the matched originGroup, we traverse use-def chains of
+    /// combinational logic, and inline assignments from the defining
+    /// combinational groups.
+    std::function<void(calyx::GroupInterface, bool)> recurseInline =
+        [&](calyx::GroupInterface recGroupOp, bool init) {
+          inlinedGroups.insert(recGroupOp);
+          for (auto assignOp :
+               recGroupOp.getBody()->getOps<calyx::AssignOp>()) {
+            if (!init) {
+              /// Inline the assignment into the originGroup.
+              auto clonedAssignOp = rewriter.clone(*assignOp.getOperation());
+              clonedAssignOp->moveBefore(originGroup.getBody(),
+                                         originGroup.getBody()->end());
+            }
+            auto src = assignOp.src();
+            auto srcDefOp = src.getDefiningOp();
+
+            /// Things which stop recursive inlining (or in other words, what
+            /// breaks combinational paths).
+            /// - Component inputs
+            /// - Register and memory reads
+            /// - Constant ops
```

Comment:
Why do constant operations stop inlining?

Reply:
Constant ops are not evaluated by any group, hence the following `state.getEvaluatingGroup(...)` would fail. I'll document this.